Nexus Tower Scripts

enteryournamehere opened the pull request:
**Paradox**
- [x] Teleporters
- [x] Darkitect cinematic
- [x] Power panels

**Assembly**
- [x] Tubes (similar to Paradox teleporters)
 
**Venture League**
- [x] Cannon (that thing to get to Hael Storm)
- [x] Speed boost pads

**Sentinels**
- [x] Sentinel speed boost pads
- [x] Armor & health dispensers
- [x] Maelstrom Infected Bricks to Faction Tokens trade point
- [x] Armory Challenge
- [x] Wishing Well
- [x] X-ray scanner thing near entrance _(client script function doesn't get called for unknown reasons; cinematic does not play and player does not move)_

**General**
- [x] Crux Prime console - also Crux Prime -> Nexus Tower
- [x] Starbase 3001 portal - also in Nimbus Station
- [x] LEGO® Club portal - also in Nimbus Station

**More things**
- [x] Scripts for [crabs](https://lu.lcdruniverse.org/explorer/objects/7000) and [Vegan Steven](https://lu.lcdruniverse.org/explorer/objects/3820) and [starfish](https://lu.lcdruniverse.org/explorer/objects/6954)

This is meant to be a complete list of NT-specific things that don't work, so please let me know if I've missed anything.

enteryournamehere committed:
Fix teleport GM rotation

enteryournamehere committed:
Paradox teleporters

enteryournamehere committed:
Assembly tubes

enteryournamehere committed:
Sentinel speed pads

enteryournamehere committed:
Sentinel area interaction things

enteryournamehere committed:
Venture cannon

enteryournamehere committed:
Venture League speed boost things

enteryournamehere committed:
Merge branch 'dev' of github.com:UchuServer/Uchu into enhancement/nexus-tower-scripts

MustacheMan2 commented:
Sentinel baggage scanner near the entrance

enteryournamehere committed:
Merge with dev (struct packets)

enteryournamehere committed:
Resolve conflicts

enteryournamehere committed:
Improve teleporters

enteryournamehere committed:
Terminate interactions

enteryournamehere committed:
Darkitect cinematic script

enteryournamehere committed:
Paradox power panels

enteryournamehere committed:
Sentinel X-ray script (only does glowing effect)

enteryournamehere committed:
Add some comments

MickVermeulen commented on Uchu.StandardScripts/NexusTower/MaelstromConverter.cs:
```diff
@@ -0,0 +1,32 @@
+            {
+                // Let the player trade 25 Maelstrom Infected Bricks for 5 faction tokens
+                var inventory = player.GetComponent<InventoryManagerComponent>();
+                inventory.RemoveLotAsync(Lot.MaelstromInfectedBrick, 25);
```
maybe add a check here if the player has 25 bricks?

enteryournamehere replied:
The client doesn't let the player interact if they have less than 25 bricks, but you're right it would probably be better to check it here as well

MickVermeulen commented on Uchu.StandardScripts/NexusTower/PowerPanel.cs:
```diff
@@ -0,0 +1,101 @@
+                        Originator = gameObject,
+                        KnockbackTime = 200,
+                        Vector = Vector3.Transform(new Vector3(15,
+                                5,
```
this formatting looks a bit odd, is this due to conflicting line endings?

enteryournamehere replied:
Not sure how that happened, thanks for pointing it out :+1:

MickVermeulen reviewed (changes requested):
Looks good, just had one discussion point

enteryournamehere committed:
Fix formatting

enteryournamehere committed:
Merge branch 'dev' of github.com:UchuServer/Uchu into enhancement/nexus-tower-scripts

enteryournamehere committed:
Combat Challenge

Known issue: countdown timer doesn't display orange time-elapsed indicator after first use of combat challenge

enteryournamehere committed:
NT world teleporters

enteryournamehere committed:
Wishing Well + multiple ScriptName attributes

enteryournamehere committed:
Ensure player has enough maelstrom infected bricks

enteryournamehere committed:
Replace insane code, support nested dictionaries.

enteryournamehere committed:
More world teleporters

enteryournamehere committed:
l_wild_ambients and l_wild_ambient_crab scripts

enteryournamehere committed:
Fix NT portal to LEGO® Club world

enteryournamehere committed:
Move OnMessageBoxRespond to associated GameObject

enteryournamehere committed:
Combat challenge: progress mission 1010

enteryournamehere committed:
Spawn targets in the correct orientation

MickVermeulen reviewed (commented):
Looks good, I did notice a lot of use of `this`, which is redundant as I've said before. No worries if this is the style you're going for now but it's not consistent with the MS styelguide afaik.

enteryournamehere committed:
Make scripts progress misison 1047

enteryournamehere commented:
> Looks good, I did notice a lot of use of `this`, which is redundant as I've said before. No worries if this is the style you're going for now but it's not consistent with the MS styelguide afaik.

It is indeed not part of the MS style guide, but as @TheNexusAvenger preferred it I've started using it as well.

enteryournamehere committed:
Improve fountain drops

enteryournamehere committed:
Merge branch 'dev' of github.com:UchuServer/Uchu into enhancement/nexus-tower-scripts

enteryournamehere committed:
Maelstrom orb: set flag 1911

enteryournamehere committed:
Speedy Traveler achievement

enteryournamehere committed:
Fix combat challenge targets not being destroyed

enteryournamehere committed:
possibly/hopefully fix combat challenge issues

enteryournamehere committed:
Handle some possible errors

TheNexusAvenger commented on Uchu.StandardScripts/NexusTower/AssemblyTube.cs:
```diff
@@ -0,0 +1,76 @@
+                    return;
+
+                // Find teleporter target location
+                var target = Zone.GameObjects.FirstOrDefault(obj =>
+                    obj.Settings.TryGetValue("groupID", out var group2)
+                    && ((string) group2).Split(";").Contains(group));
```
Any reason `GetGroup` isn't being used?
https://github.com/UchuServer/Uchu/blob/d3e91e4f3c002ff181567814bf8af7e4e4671c5a/Uchu.World/Scripting/Native/NativeScript.cs#L81

enteryournamehere replied:
Nope, I just wasn't aware of that method. Thanks!

TheNexusAvenger commented on Uchu.World/Objects/GameObjects/Player.cs:
```diff
@@ -374,12 +371,14 @@ private void UpdatePhysics(Vector3 position, Quaternion rotation)
                 Associate = this,
+                SetRotation = rotation != null,
                 Position = position,
+                Rotation = rotation?? Quaternion.Identity,
```
Minor formatting thing: should this be `rotation ?? Quaternion.Identity,` with another space?

enteryournamehere replied:
Yes, thanks

TheNexusAvenger commented on Uchu.World/Objects/GameObjects/GameObject.cs:
```diff
@@ -106,6 +106,8 @@ public int GameMasterLevel
 
         public Event OnRailMovementReady { get; }
 
+        public Event<Player, MessageBoxRespondMessage> OnMessageBoxRespond { get; }
```
(No change required)
Interesting that this event was moved here. Guessing that this reduces the connections required.

enteryournamehere replied:
Previously the associated game object was discarded, which makes it impossible to distinguish between message boxes of different world teleporters (Crux Prime console, Starbase portal, etc). By moving the event here you also don't need wrap the message box response listener in a `Listen(Zone.OnPlayerLoad){}`.

TheNexusAvenger commented on Uchu.World/Handlers/GameMessages/GeneralHandler.cs:
```diff
@@ -176,7 +176,7 @@ public void RequestPlatformResyncHandler(RequestPlatformResyncMessage message, P
         public void MessageBoxRespondMessageHandler(MessageBoxRespondMessage message, Player player)
         {
-            player.OnMessageBoxRespond.Invoke(message.Button, message.Identifier, message.UserData);
+            message.Associate.OnMessageBoxRespond.Invoke(player, message);
```
Any chance `message.Associate` will be `null` and throw a null exception?

enteryournamehere replied:
I've added a null-conditional operator `?.` to be safe.

TheNexusAvenger commented on Uchu.StandardScripts/NexusTower/ParadoxTeleporter.cs:
```diff
@@ -0,0 +1,93 @@
+                    return;
+
+                // Find teleporter target location.
+                var target = Zone.GameObjects.FirstOrDefault(obj => obj.Settings.TryGetValue("groupID", out var group2)
+                                                       && ((string) group2).Split(";").Contains(group));
```
Same comment as the Assembly teleporter with `GetGroup`.

TheNexusAvenger reviewed (commented):
As usual, great set of improvements. Got a few comments.

enteryournamehere committed:
Use GetGroup method

enteryournamehere committed:
Formatting

enteryournamehere committed:
Avoid possible null reference exception
